README.md
# secp256k1-py ![Build Status](https://github.com/rustyrussell/secp256k1-py/actions/workflows/python-app.yml/badge.svg)

Python FFI bindings for [libsecp256k1](https://github.com/bitcoin/secp256k1)
(an experimental and optimized C library for EC operations on curve secp256k1).

Previously maintained by Ludvig Broberg, now at https://github.com/rustyrussell/secp256k1-py .

## Installation

```
pip install secp256k1
```

### Precompiled binary packages (wheels)

Precompiled binary wheels are available on Linux.

In case you don't want to use the binary packages you can prevent pip from
using them with the following command:

```
pip install --no-binary :all: secp256k1
```


### Installation with compilation

If you either can't or don't want to use the binary package options described
above read on to learn what is needed to install the source pacakge.

The library bundles its own libsecp256k1 currently, as there is no
versioning to allow us to safely determine compatibility with an
installed library, especially as we also build all the experimental
modules.

For the bundled version to compile successfully you need to have a C compiler
as well as the development headers for `libffi` and `libgmp` installed.

On Debian / Ubuntu for example the necessary packages are:

* `build-essential`
* `automake`
* `pkg-config`
* `libtool`
* `libffi-dev`

On OS X the necessary homebrew packages are:

* `automake`
* `pkg-config`
* `libtool`
* `libffi`


## Command line usage

###### Generate a private key and show the corresponding public key

```
$ python -m secp256k1 privkey -p

a1455c78a922c52f391c5784f8ca1457367fa57f9d7a74fdab7d2c90ca05c02e
Public key: 02477ce3b986ab14d123d6c4167b085f4d08c1569963a0201b2ffc7d9d6086d2f3
```

###### Sign a message

```
$ python -m secp256k1 sign \
	-k a1455c78a922c52f391c5784f8ca1457367fa57f9d7a74fdab7d2c90ca05c02e \
	-m hello

3045022100a71d86190354d64e5b3eb2bd656313422cdf7def69bf3669cdbfd09a9162c96e0220713b81f3440bff0b639d2f29b2c48494b812fa89b754b7b6cdc9eaa8027cf369
```

###### Check signature

```
$ python -m secp256k1 checksig \
	-p 02477ce3b986ab14d123d6c4167b085f4d08c1569963a0201b2ffc7d9d6086d2f3 \
	-m hello \
	-s 3045022100a71d86190354d64e5b3eb2bd656313422cdf7def69bf3669cdbfd09a9162c96e0220713b81f3440bff0b639d2f29b2c48494b812fa89b754b7b6cdc9eaa8027cf369

True
```

###### Generate a signature that allows recovering the public key

```
$ python -m secp256k1 signrec \
	-k a1455c78a922c52f391c5784f8ca1457367fa57f9d7a74fdab7d2c90ca05c02e \
	-m hello

515fe95d0780b11633f3352deb064f1517d58f295a99131e9389da8bfacd64422513d0cd4e18a58d9f4873b592afe54cf63e8f294351d1e612c8a297b5255079 1
```

###### Recover public key

```
$ python -m secp256k1 recpub \
	-s 515fe95d0780b11633f3352deb064f1517d58f295a99131e9389da8bfacd64422513d0cd4e18a58d9f4873b592afe54cf63e8f294351d1e612c8a297b5255079 \
	-i 1 \
	-m hello

Public key: 02477ce3b986ab14d123d6c4167b085f4d08c1569963a0201b2ffc7d9d6086d2f3
```


It is easier to get started with command line, but it is more common to use this as a library. For that, check the next sections.


## API

#### class `secp256k1.PrivateKey(privkey, raw)`

The `PrivateKey` class loads or creates a private key by obtaining 32 bytes from urandom and operates over it.

##### Instantiation parameters

- `privkey=None` - generate a new private key if None, otherwise load a private key.
- `raw=True` - if `True`, it is assumed that `privkey` is just a sequence of bytes, otherwise it is assumed that it is in the DER format. This is not used when `privkey` is not specified.

##### Methods and instance attributes

- `pubkey`: an instance of `secp256k1.PublicKey`.
- `private_key`: raw bytes for the private key.

- `set_raw_privkey(privkey)`<br/>
update the `private_key` for this instance with the bytes specified by `privkey`. If `privkey` is invalid, an Exception is raised. The `pubkey` is also updated based on the new private key.

- `serialize()` -> bytes<br/>
convert the raw bytes present in `private key` to a hexadecimal string.

- `deserialize(privkey_ser)` -> bytes<br/>
convert from a hexadecimal string to raw bytes and update the `pubkey` and `private_key` for this instance.

- `tweak_add(scalar)` -> bytes<br/>
tweak the current private key by adding a 32 byte scalar to it and return a new raw private key composed of 32 bytes.

- `tweak_mul(scalar)` -> bytes<br/>
tweak the current private key by multiplying it by a 32 byte scalar and return a new raw private key composed of 32 bytes.

- `ecdsa_sign(msg, raw=False, digest=hashlib.sha256)` -> internal object<br/>
by default, create an ECDSA-SHA256 signature from the bytes in `msg`. If `raw` is True, then the `digest` function is not applied over `msg`, otherwise the `digest` must produce 256 bits or an `Exception` will be raised.<br/><br/>
The returned object is a structure from the C lib. If you want to store it (on a disk or similar), use `ecdsa_serialize` and later on use `ecdsa_deserialize` when loading.

- `ecdsa_sign_recoverable(msg, raw=False, digest=hashlib.sha256)` -> internal object<br/>
create a recoverable ECDSA signature. See `ecdsa_sign` for parameters description.

- `schnorr_sign(msg, bip340tag, raw=False)` -> bytes<br/>

create a BIP-340 signature for `msg`; `bip340tag` should be a string
or byte value which distinguishes this usage from any other usage of
signatures (e.g. your program name, or full protocol name).  If `raw`
is specified, then `bip340tag` is not used, and the `msg` (usually
a 32-byte hash) is signed directly.

It produces non-malleable 64-byte signatures which support batch
validation.


#### class `secp256k1.PublicKey(pubkey, raw)`

The `PublicKey` class loads an existing public key and operates over it.

##### Instantiation parameters

- `pubkey=None` - do not load a public key if None, otherwise do.
- `raw=False` - if `False`, it is assumed that `pubkey` has gone through `PublicKey.deserialize` already, otherwise it must be specified as bytes.

##### Methods and instance attributes

- `public_key`: an internal object representing the public key.

- `serialize(compressed=True)` -> bytes<br/>
convert the `public_key` to bytes. If `compressed` is True, 33 bytes will be produced, otherwise 65 will be.

- `deserialize(pubkey_ser)` -> internal object<br/>
convert the bytes resulting from a previous `serialize` call back to an internal object and update the `public_key` for this instance. The length of `pubkey_ser` determines if it was serialized with `compressed=True` or not. This will raise an Exception if the size is invalid or if the key is invalid.

- `combine(pubkeys)` -> internal object<br/>
combine multiple public keys (those returned from `PublicKey.deserialize`) and return a public key (which can be serialized as any other regular public key). The `public_key` for this instance is updated to use the resulting combined key. If it is not possible the combine the keys, an Exception is raised.

- `tweak_add(scalar)` -> internal object<br/>
tweak the current public key by adding a 32 byte scalar times the generator to it and return a new PublicKey instance.

- `tweak_mul(scalar)` -> internal object<br/>
tweak the current public key by multiplying it by a 32 byte scalar and return a new PublicKey instance.

- `ecdsa_verify(msg, raw_sig, raw=False, digest=hashlib.sha256)` -> bool<br/>
verify an ECDSA signature and return True if the signature is correct, False otherwise. `raw_sig` is expected to be an object returned from `ecdsa_sign` (or if it was serialized using `ecdsa_serialize`, then first run it through `ecdsa_deserialize`). `msg`, `raw`, and `digest` are used as described in `ecdsa_sign`.

- `schnorr_verify(msg, schnorr_sig, bip340tag, raw=False)` -> bool<br/>
verify a Schnorr signature and return True if the signature is correct, False otherwise. `schnorr_sig` is expected to be the result from `schnorr_sign`, `msg`, `bip340tag` and `raw` must match those used in `schnorr_sign`.

- `ecdh(scalar, hashfn=ffi.NULL, hasharg=ffi.NULL)` -> bytes<br/>
compute an EC Diffie-Hellman secret in constant time. The instance `public_key` is used as the public point, and the `scalar` specified must be composed of 32 bytes. It outputs 32 bytes representing the ECDH secret computed. The hashing function can be overridden, but (unlike libsecp256k1 itself) we insist that it produce 32-bytes of output. If the `scalar` is invalid, an Exception is raised.


#### class `secp256k1.ECDSA`

The `ECDSA` class is intended to be used as a mix in. Its methods can be accessed from any `secp256k1.PrivateKey` or `secp256k1.PublicKey` instances.

##### Methods

- `ecdsa_serialize(raw_sig)` -> bytes<br/>
convert the result from `ecdsa_sign` to DER.

- `ecdsa_deserialie(ser_sig)` -> internal object<br/>
convert DER bytes to an internal object.

- `ecdsa_serialize_compact(raw_sig)` -> bytes<br/>
convert the result from `ecdsa_sign` to a compact serialization of 64 bytes.

- `ecdsa_deserialize_compact(ser_sig)` -> internal object<br/>
convert a compact serialization of 64 bytes to an internal object.

- `ecdsa_signature_normalize(raw_sig, check_only=False)` -> (bool, internal object | None)<br/>
check and optionally convert a signature to a normalized lower-S form. If `check_only` is True then the normalized signature is not returned.<br/><br/>
This function always return a tuple containing a boolean (True if not previously normalized or False if signature was already normalized), and the normalized signature. When `check_only` is True, the normalized signature returned is always None.

- `ecdsa_recover(msg, recover_sig, raw=False, digest=hashlib.sha256)` -> internal object<br/>
recover an ECDSA public key from a signature generated by `ecdsa_sign_recoverable`. `recover_sig` is expected to be an object returned from `ecdsa_sign_recoverable` (or if it was serialized using `ecdsa_recoverable_serialize`, then first run it through `ecdsa_recoverable_deserialize`). `msg`, `raw`, and `digest` are used as described in `ecdsa_sign`.<br/><br/>

- `ecdsa_recoverable_serialize(recover_sig)` -> (bytes, int)<br/>
convert the result from `ecdsa_sign_recoverable` to a tuple composed of 65 bytesand an integer denominated as recovery id.

- `ecdsa_recoverable_deserialize(ser_sig, rec_id)`-> internal object<br/>
convert the result from `ecdsa_recoverable_serialize` back to an internal object that can be used by `ecdsa_recover`.

- `ecdsa_recoverable_convert(recover_sig)` -> internal object<br/>
convert a recoverable signature to a normal signature, i.e. one that can be used by `ecdsa_serialize` and related methods.


## Example

```python
from secp256k1 import PrivateKey, PublicKey

privkey = PrivateKey()
privkey_der = privkey.serialize()
assert privkey.deserialize(privkey_der) == privkey.private_key

sig = privkey.ecdsa_sign(b'hello')
verified = privkey.pubkey.ecdsa_verify(b'hello', sig)
assert verified

sig_der = privkey.ecdsa_serialize(sig)
sig2 = privkey.ecdsa_deserialize(sig_der)
vrf2 = privkey.pubkey.ecdsa_verify(b'hello', sig2)
assert vrf2

pubkey = privkey.pubkey
pub = pubkey.serialize()

pubkey2 = PublicKey(pub, raw=True)
assert pubkey2.serialize() == pub
assert pubkey2.ecdsa_verify(b'hello', sig)
```

```python
from secp256k1 import PrivateKey

key = '31a84594060e103f5a63eb742bd46cf5f5900d8406e2726dedfc61c7cf43ebad'
msg = '9e5755ec2f328cc8635a55415d0e9a09c2b6f2c9b0343c945fbbfe08247a4cbe'
sig = '30440220132382ca59240c2e14ee7ff61d90fc63276325f4cbe8169fc53ade4a407c2fc802204d86fbe3bde6975dd5a91fdc95ad6544dcdf0dab206f02224ce7e2b151bd82ab'

privkey = PrivateKey(bytes(bytearray.fromhex(key)), raw=True)
sig_check = privkey.ecdsa_sign(bytes(bytearray.fromhex(msg)), raw=True)
sig_ser = privkey.ecdsa_serialize(sig_check)

assert sig_ser == bytes(bytearray.fromhex(sig))
```

```python
from secp256k1 import PrivateKey

key = '7ccca75d019dbae79ac4266501578684ee64eeb3c9212105f7a3bdc0ddb0f27e'
pub_compressed = '03e9a06e539d6bf5cf1ca5c41b59121fa3df07a338322405a312c67b6349a707e9'
pub_uncompressed = '04e9a06e539d6bf5cf1ca5c41b59121fa3df07a338322405a312c67b6349a707e94c181c5fe89306493dd5677143a329065606740ee58b873e01642228a09ecf9d'

privkey = PrivateKey(bytes(bytearray.fromhex(key)))
pubkey_ser = privkey.pubkey.serialize()
pubkey_ser_uncompressed = privkey.pubkey.serialize(compressed=False)

assert pubkey_ser == bytes(bytearray.fromhex(pub_compressed))
assert pubkey_ser_uncompressed == bytes(bytearray.fromhex(pub_uncompressed))
```


## Technical details about the bundled libsecp256k1

The bundling of libsecp256k1 is handled by the various setup.py build phases:

- During 'sdist':
  If the directory `libsecp256k1` doesn't exist in the
  source directory it is downloaded from the location specified
  by the `LIB_TARBALL_URL` constant in `setup.py` and extracted into
  a directory called `libsecp256k1`

  To upgrade to a newer version of the bundled libsecp256k1 source
  simply delete the `libsecp256k1` directory and update the
  `LIB_TARBALL_URL` to point to a newer commit.

- During 'install':
  To support (future) use of system libsecp256k1, and because of the
  way the way cffi modules are implemented it is necessary
  to perform system library detection in the cffi build module
  `_cffi_build/build.py` as well as in `setup.py`. For that reason
  some utility functions have been moved into a `setup_support.py`
  module which is imported from both.

  By default, the bundled source code is used to build a library
  locally that will be statically linked into the CFFI extension.

  You can set the environment variable `SECP_BUNDLED_NO_EXPERIMENTAL`
  to disable all experimental modules except the `recovery` module.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.14.0 - 2021-10-19: "In Which Bitcoin Twitter Rescues Rusty"

After 5 years of neglect, I filed a request to take over maintenance.
I hope the previous author (Ludvig Broberg) is living a joyous life
somewhere, and if he ever looks at this, is proud of the project he
created!

### Added

- extrakeys interfaces (still needs documentation).
- secp256k1_tagged_sha256 function exposed via FFI.
- CHANGELOG.md started

### Changed

- Latest libsecp256k1 version used.
- Schnorr interface is completely changed to reflect the underlying interface
  changes (which follow [BIP-340]).
- ECDH interface now supports using your own hashing functions.
- Uses a global libsecp256k1 context.
- Rusty Russell is now self-appointed maintainer.

### Removed

- Schnorr partial and pair functions removed (removed from underlying lib)
- Requirement for gmp removed (removed from underlying lib)
- Support for Python 2
- Support for supplying your own secp256k1 context per PublicKey/PrivateKey

### Fixed

- Bundled source fetch now works (old git reference URL was now 404).
- Build had rotted, fixes implemented (not sure how it ever worked?).

[BIP-340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki


tests/test_ecdh.py
import pytest

import secp256k1


def test_ecdh():
    if not secp256k1.HAS_ECDH:
        pytest.skip("secp256k1_ecdh not enabled, skipping")
        return

    pubkey = secp256k1.PrivateKey().pubkey

    p = secp256k1.PublicKey(pubkey.public_key)
    with pytest.raises(Exception):
        # Bad scalar length.
        p.ecdh(b"")
    with pytest.raises(Exception):
        # Bad scalar type.
        p.ecdh([])

    res = p.ecdh(b"0" * 32)
    assert type(res) == bytes


secp256k1/__init__.py
import binascii
import hashlib
import os

from ._libsecp256k1 import ffi, lib

EC_COMPRESSED = lib.SECP256K1_EC_COMPRESSED
EC_UNCOMPRESSED = lib.SECP256K1_EC_UNCOMPRESSED

HAS_RECOVERABLE = hasattr(lib, "secp256k1_ecdsa_sign_recoverable")
HAS_SCHNORR = hasattr(lib, "secp256k1_schnorrsig_sign")
HAS_ECDH = hasattr(lib, "secp256k1_ecdh")
HAS_EXTRAKEYS = hasattr(lib, "secp256k1_keypair_create")

# Keeping a single one of these is most efficient.
secp256k1_ctx = lib.secp256k1_context_create(
    lib.SECP256K1_CONTEXT_SIGN | lib.SECP256K1_CONTEXT_VERIFY
)


class ECDSA:
    def ecdsa_serialize(self, raw_sig):
        len_sig = 74
        output = ffi.new("unsigned char[%d]" % len_sig)
        outputlen = ffi.new("size_t *", len_sig)

        res = lib.secp256k1_ecdsa_signature_serialize_der(
            secp256k1_ctx, output, outputlen, raw_sig
        )
        assert res == 1

        return bytes(ffi.buffer(output, outputlen[0]))

    def ecdsa_deserialize(self, ser_sig):
        raw_sig = ffi.new("secp256k1_ecdsa_signature *")
        res = lib.secp256k1_ecdsa_signature_parse_der(
            secp256k1_ctx, raw_sig, ser_sig, len(ser_sig)
        )
        assert res == 1

        return raw_sig

    def ecdsa_serialize_compact(self, raw_sig):
        len_sig = 64
        output = ffi.new("unsigned char[%d]" % len_sig)

        res = lib.secp256k1_ecdsa_signature_serialize_compact(
            secp256k1_ctx, output, raw_sig
        )
        assert res == 1

        return bytes(ffi.buffer(output, len_sig))

    def ecdsa_deserialize_compact(self, ser_sig):
        if len(ser_sig) != 64:
            raise Exception("invalid signature length")

        raw_sig = ffi.new("secp256k1_ecdsa_signature *")
        res = lib.secp256k1_ecdsa_signature_parse_compact(
            secp256k1_ctx, raw_sig, ser_sig
        )
        assert res == 1

        return raw_sig

    def ecdsa_signature_normalize(self, raw_sig, check_only=False):
        """
        Check and optionally convert a signature to a normalized lower-S form.
        If check_only is True then the normalized signature is not returned.

        This function always return a tuple containing a boolean (True if
        not previously normalized or False if signature was already
        normalized), and the normalized signature. When check_only is True,
        the normalized signature returned is always None.
        """
        if check_only:
            sigout = ffi.NULL
        else:
            sigout = ffi.new("secp256k1_ecdsa_signature *")

        result = lib.secp256k1_ecdsa_signature_normalize(secp256k1_ctx, sigout, raw_sig)

        return (bool(result), sigout if sigout != ffi.NULL else None)

    def ecdsa_recover(self, msg, recover_sig, raw=False, digest=hashlib.sha256):
        if not HAS_RECOVERABLE:
            raise Exception("secp256k1_recovery not enabled")

        msg32 = _hash32(msg, raw, digest)
        pubkey = ffi.new("secp256k1_pubkey *")

        recovered = lib.secp256k1_ecdsa_recover(
            secp256k1_ctx, pubkey, recover_sig, msg32
        )
        if recovered:
            return pubkey
        raise Exception("failed to recover ECDSA public key")

    def ecdsa_recoverable_serialize(self, recover_sig):
        if not HAS_RECOVERABLE:
            raise Exception("secp256k1_recovery not enabled")

        outputlen = 64
        output = ffi.new("unsigned char[%d]" % outputlen)
        recid = ffi.new("int *")

        lib.secp256k1_ecdsa_recoverable_signature_serialize_compact(
            secp256k1_ctx, output, recid, recover_sig
        )

        return bytes(ffi.buffer(output, outputlen)), recid[0]

    def ecdsa_recoverable_deserialize(self, ser_sig, rec_id):
        if not HAS_RECOVERABLE:
            raise Exception("secp256k1_recovery not enabled")
        if rec_id < 0 or rec_id > 3:
            raise Exception("invalid rec_id")
        if len(ser_sig) != 64:
            raise Exception("invalid signature length")

        recover_sig = ffi.new("secp256k1_ecdsa_recoverable_signature *")

        parsed = lib.secp256k1_ecdsa_recoverable_signature_parse_compact(
            secp256k1_ctx, recover_sig, ser_sig, rec_id
        )
        if parsed:
            return recover_sig
        else:
            raise Exception("failed to parse ECDSA compact sig")

    def ecdsa_recoverable_convert(self, recover_sig):
        if not HAS_RECOVERABLE:
            raise Exception("secp256k1_recovery not enabled")

        normal_sig = ffi.new("secp256k1_ecdsa_signature *")

        lib.secp256k1_ecdsa_recoverable_signature_convert(
            secp256k1_ctx, normal_sig, recover_sig
        )

        return normal_sig


class PublicKey(ECDSA):
    def __init__(self, pubkey=None, raw=False):
        if pubkey is not None:
            if raw:
                if not isinstance(pubkey, bytes):
                    raise TypeError("raw pubkey must be bytes")
                self.public_key = self.deserialize(pubkey)
            else:
                if not isinstance(pubkey, ffi.CData):
                    raise TypeError("pubkey must be an internal object")
                assert ffi.typeof(pubkey) is ffi.typeof("secp256k1_pubkey *")
                self.public_key = pubkey
            self._pubkey_changed()
        else:
            self.public_key = None

    def _pubkey_changed(self):
        if HAS_EXTRAKEYS:
            self.xonly_pubkey = ffi.new("secp256k1_xonly_pubkey *")
            assert (
                lib.secp256k1_xonly_pubkey_from_pubkey(
                    secp256k1_ctx, self.xonly_pubkey, ffi.NULL, self.public_key
                )
                == 1
            )

    def serialize(self, compressed=True):
        assert self.public_key, "No public key defined"

        len_compressed = 33 if compressed else 65
        res_compressed = ffi.new("char [%d]" % len_compressed)
        outlen = ffi.new("size_t *", len_compressed)
        compflag = EC_COMPRESSED if compressed else EC_UNCOMPRESSED

        serialized = lib.secp256k1_ec_pubkey_serialize(
            secp256k1_ctx, res_compressed, outlen, self.public_key, compflag
        )
        assert serialized == 1

        return bytes(ffi.buffer(res_compressed, len_compressed))

    def deserialize(self, pubkey_ser):
        if len(pubkey_ser) not in (33, 65):
            raise Exception("unknown public key size (expected 33 or 65)")

        pubkey = ffi.new("secp256k1_pubkey *")

        res = lib.secp256k1_ec_pubkey_parse(
            secp256k1_ctx, pubkey, pubkey_ser, len(pubkey_ser)
        )
        if not res:
            raise Exception("invalid public key")

        self.public_key = pubkey
        self._pubkey_changed()
        return pubkey

    def combine(self, pubkeys):
        """Add a number of public keys together."""
        assert len(pubkeys) > 0

        outpub = ffi.new("secp256k1_pubkey *")
        for item in pubkeys:
            assert ffi.typeof(item) is ffi.typeof("secp256k1_pubkey *")

        res = lib.secp256k1_ec_pubkey_combine(
            secp256k1_ctx, outpub, pubkeys, len(pubkeys)
        )
        if not res:
            raise Exception("failed to combine public keys")

        self.public_key = outpub
        self._pubkey_changed()
        return outpub

    def tweak_add(self, scalar):
        """
        Tweak the current public key by adding a 32 byte scalar times
        the generator to it and return a new PublicKey instance.
        """
        return _tweak_public(self, lib.secp256k1_ec_pubkey_tweak_add, scalar)

    def tweak_mul(self, scalar):
        """
        Tweak the current public key by multiplying it by a 32 byte scalar
        and return a new PublicKey instance.
        """
        return _tweak_public(self, lib.secp256k1_ec_pubkey_tweak_mul, scalar)

    def ecdsa_verify(self, msg, raw_sig, raw=False, digest=hashlib.sha256):
        assert self.public_key, "No public key defined"

        msg32 = _hash32(msg, raw, digest)

        verified = lib.secp256k1_ecdsa_verify(
            secp256k1_ctx, raw_sig, msg32, self.public_key
        )

        return bool(verified)

    def schnorr_verify(self, msg, schnorr_sig, bip340tag, raw=False):
        assert self.public_key, "No public key defined"
        if not HAS_SCHNORR:
            raise Exception("secp256k1_schnorr not enabled")

        msg_to_sign = _bip340_tag(msg, raw, bip340tag)

        verified = lib.secp256k1_schnorrsig_verify(
            secp256k1_ctx, schnorr_sig, msg_to_sign, len(msg_to_sign), self.xonly_pubkey
        )

        return bool(verified)

    def ecdh(self, scalar, hashfn=ffi.NULL, hasharg=ffi.NULL):
        assert self.public_key, "No public key defined"
        if not HAS_ECDH:
            raise Exception("secp256k1_ecdh not enabled")
        # Technically, it need only match the hashfn, but this is standard.
        if not isinstance(scalar, bytes) or len(scalar) != 32:
            raise TypeError("scalar must be composed of 32 bytes")

        result = ffi.new("char [32]")

        res = lib.secp256k1_ecdh(
            secp256k1_ctx, result, self.public_key, scalar, hashfn, hasharg
        )
        if not res:
            raise Exception(f"invalid scalar ({res})")

        return bytes(ffi.buffer(result, 32))


class PrivateKey(ECDSA):
    def __init__(self, privkey=None, raw=True):
        self.pubkey = None
        self.private_key = None
        if privkey is None:
            self.set_raw_privkey(_gen_private_key())
        else:
            if raw:
                if not isinstance(privkey, bytes) or len(privkey) != 32:
                    raise TypeError("privkey must be composed of 32 bytes")
                self.set_raw_privkey(privkey)
            else:
                self.deserialize(privkey)

    def _update_public_key(self):
        public_key = self._gen_public_key(self.private_key)
        self.pubkey = PublicKey(public_key, raw=False)
        if HAS_EXTRAKEYS:
            self.keypair = ffi.new("secp256k1_keypair *")
            if (
                lib.secp256k1_keypair_create(
                    secp256k1_ctx, self.keypair, self.private_key
                )
                != 1
            ):
                raise Exception("invalid private key (can't make keypair?)")

    def set_raw_privkey(self, privkey):
        if not lib.secp256k1_ec_seckey_verify(secp256k1_ctx, privkey):
            raise Exception("invalid private key")
        self.private_key = privkey
        self._update_public_key()

    def serialize(self):
        hexkey = binascii.hexlify(self.private_key)
        return hexkey.decode("utf8")

    def deserialize(self, privkey_ser):
        if len(privkey_ser) != 64:
            raise Exception("invalid private key")
        rawkey = binascii.unhexlify(privkey_ser)

        self.set_raw_privkey(rawkey)
        return self.private_key

    def _gen_public_key(self, privkey):
        pubkey_ptr = ffi.new("secp256k1_pubkey *")

        created = lib.secp256k1_ec_pubkey_create(secp256k1_ctx, pubkey_ptr, privkey)
        assert created == 1

        return pubkey_ptr

    def tweak_add(self, scalar):
        """
        Tweak the current private key by adding a 32 byte scalar
        to it and return a new raw private key composed of 32 bytes.
        """
        return _tweak_private(self, lib.secp256k1_ec_privkey_tweak_add, scalar)

    def tweak_mul(self, scalar):
        """
        Tweak the current private key by multiplying it by a 32 byte scalar
        and return a new raw private key composed of 32 bytes.
        """
        return _tweak_private(self, lib.secp256k1_ec_privkey_tweak_mul, scalar)

    def ecdsa_sign(self, msg, raw=False, digest=hashlib.sha256, custom_nonce=None):
        msg32 = _hash32(msg, raw, digest)
        raw_sig = ffi.new("secp256k1_ecdsa_signature *")
        nonce_fn = ffi.NULL
        nonce_data = ffi.NULL
        if custom_nonce:
            nonce_fn, nonce_data = custom_nonce
        signed = lib.secp256k1_ecdsa_sign(
            secp256k1_ctx, raw_sig, msg32, self.private_key, nonce_fn, nonce_data
        )
        assert signed == 1

        return raw_sig

    def ecdsa_sign_recoverable(self, msg, raw=False, digest=hashlib.sha256):
        if not HAS_RECOVERABLE:
            raise Exception("secp256k1_recovery not enabled")

        msg32 = _hash32(msg, raw, digest)
        raw_sig = ffi.new("secp256k1_ecdsa_recoverable_signature *")

        signed = lib.secp256k1_ecdsa_sign_recoverable(
            secp256k1_ctx, raw_sig, msg32, self.private_key, ffi.NULL, ffi.NULL
        )
        assert signed == 1

        return raw_sig

    def schnorr_sign(self, msg, bip340tag, raw=False):
        if not HAS_SCHNORR:
            raise Exception("secp256k1_schnorr not enabled")

        msg_to_sign = _bip340_tag(msg, raw, bip340tag)
        sig64 = ffi.new("char [64]")

        # FIXME: It's recommended to provide aux_rand32...
        signed = lib.secp256k1_schnorrsig_sign_custom(
            secp256k1_ctx, sig64, msg_to_sign, len(msg_to_sign), self.keypair, ffi.NULL
        )
        assert signed == 1

        return bytes(ffi.buffer(sig64, 64))


def _bip340_tag(msg, raw, tag):
    if raw:
        return msg

    if isinstance(tag, bytes):
        bytestag = tag
    else:
        bytestag = tag.encode()

    hash32 = ffi.new("char [32]")
    lib.secp256k1_tagged_sha256(
        secp256k1_ctx, hash32, bytestag, len(bytestag), msg, len(msg)
    )
    return bytes(ffi.buffer(hash32, 32))


def _hash32(msg, raw, digest):
    if not raw:
        msg32 = digest(msg).digest()
    else:
        msg32 = msg
    if len(msg32) * 8 != 256:
        raise Exception("digest function must produce 256 bits")
    return msg32


def _gen_private_key():
    key = os.urandom(32)
    return key


def _tweak_public(inst, func, scalar):
    if not isinstance(scalar, bytes) or len(scalar) != 32:
        raise TypeError("scalar must be composed of 32 bytes")
    assert inst.public_key, "No public key defined."

    # Create a copy of the current public key.
    newpub = PublicKey(inst.serialize(), raw=True)

    res = func(secp256k1_ctx, newpub.public_key, scalar)
    if not res:
        raise Exception("Tweak is out of range")

    return newpub


def _tweak_private(inst, func, scalar):
    if not isinstance(scalar, bytes) or len(scalar) != 32:
        raise TypeError("scalar must be composed of 32 bytes")

    # Create a copy of the current private key.
    key = ffi.new("char [32]", inst.private_key)

    res = func(secp256k1_ctx, key, scalar)
    if not res:
        raise Exception("Tweak is out of range")

    return bytes(ffi.buffer(key, 32))


# Apparently flake8 thinks this is "too complex".  Maybe FIXME?
def _main_cli(args, out, encoding="utf-8"):  # noqa: C901
    import binascii

    def show_public(public_key):
        rawp = public_key.serialize()
        out.write(f"Public key: {binascii.hexlify(rawp).decode(encoding)}\n")

    def sign(funcname, params):
        raw = bytes(bytearray.fromhex(params.private_key))
        priv = PrivateKey(raw)
        func = getattr(priv, funcname)
        sig = func(params.message)
        return priv, sig

    if args.action == "privkey":
        if args.private_key:
            rawkey = bytes(bytearray.fromhex(args.private_key))
        else:
            rawkey = None
        priv = PrivateKey(rawkey)
        raw = priv.private_key
        out.write(f"{binascii.hexlify(raw).decode(encoding)}\n")
        if args.show_pubkey:
            show_public(priv.pubkey)

    elif args.action == "sign":
        priv, sig_raw = sign("ecdsa_sign", args)
        sig = priv.ecdsa_serialize(sig_raw)
        out.write(f"{binascii.hexlify(sig).decode(encoding)}\n")
        if args.show_pubkey:
            show_public(priv.pubkey)

    elif args.action == "checksig":
        raw = bytes(bytearray.fromhex(args.public_key))
        sig = bytes(bytearray.fromhex(args.signature))
        pub = PublicKey(raw, raw=True)
        try:
            sig_raw = pub.ecdsa_deserialize(sig)
            good = pub.ecdsa_verify(args.message, sig_raw)
        except:  # noqa: E722
            good = False
        out.write(f"{good}\n")
        return 0 if good else 1

    elif args.action == "signrec":
        priv, sig = sign("ecdsa_sign_recoverable", args)
        sig, recid = priv.ecdsa_recoverable_serialize(sig)
        out.write(f"{binascii.hexlify(sig).decode(encoding)} {recid}\n")
        if args.show_pubkey:
            show_public(priv.pubkey)

    elif args.action == "recpub":
        empty = PublicKey()
        sig_raw = bytes(bytearray.fromhex(args.signature))
        sig = empty.ecdsa_recoverable_deserialize(sig_raw, args.recid)
        pubkey = empty.ecdsa_recover(args.message, sig)
        show_public(PublicKey(pubkey))

    return 0


def _parse_cli():
    import sys
    from argparse import ArgumentParser

    py2 = sys.version_info.major == 2
    enc = sys.getfilesystemencoding()

    def bytes_input(s):
        return s if py2 else s.encode(enc)

    parser = ArgumentParser(prog="secp256k1")
    subparser = parser.add_subparsers(dest="action")

    genparser = subparser.add_parser("privkey")
    genparser.add_argument("-p", "--show-pubkey", action="store_true")
    genparser.add_argument("-k", "--private_key")

    sign = subparser.add_parser("sign")
    sign.add_argument("-k", "--private-key", required=True)
    sign.add_argument("-m", "--message", required=True, type=bytes_input)
    sign.add_argument("-p", "--show-pubkey", action="store_true")

    signrec = subparser.add_parser("signrec")
    signrec.add_argument("-k", "--private-key", required=True)
    signrec.add_argument("-m", "--message", required=True, type=bytes_input)
    signrec.add_argument("-p", "--show-pubkey", action="store_true")

    check = subparser.add_parser("checksig")
    check.add_argument("-p", "--public-key", required=True)
    check.add_argument("-m", "--message", required=True, type=bytes_input)
    check.add_argument("-s", "--signature", required=True)

    recpub = subparser.add_parser("recpub")
    recpub.add_argument("-m", "--message", required=True, type=bytes_input)
    recpub.add_argument("-i", "--recid", required=True, type=int)
    recpub.add_argument("-s", "--signature", required=True)

    return parser, enc


def main():
    import sys

    parser, enc = _parse_cli()
    args = parser.parse_args(sys.argv[1:])
    sys.exit(_main_cli(args, sys.stdout, enc))


if __name__ == "__main__":
    main()
